### PR TITLE
Rename DataSource -> DataConnector, Update principles, add extensibility doc

### PR DESCRIPTION
**Pull request by phillipleblanc**
Renames `DataSource` to be `DataConnector` to reflect its new role in writing data back to the origin.

Adds EXTENSIBILITY.md to document the interfaces that can be extended in the Spice runtime.

Currently just documents the data extensibility points - models will be added in a follow-up PR.

### CODE REVIEW

File: docs/PRINCIPLES.md [hunk 1]
```diff
@@ -6,9 +6,7 @@ Spice.ai is built upon a foundation of first principles.
 
 - Developer experience first
 - Bring data and ML to your application
-- API first
```
Comment (lukekim) [inline docs/PRINCIPLES.md:9]: API First should remain as a principle.

File: docs/PRINCIPLES.md [hunk 2]
```diff
@@ -6,9 +6,7 @@ Spice.ai is built upon a foundation of first principles.
...
-- Composable from community driven-components
```
Comment (lukekim) [inline docs/PRINCIPLES.md:10]: Composable should probably also remain. We still have Datasets, Models, and will later have Functions.

**Review by sgrebnov — APPROVED** [review]

File: docs/EXTENSIBILITY.md [hunk 3]
```diff
@@ -0,0 +1,83 @@
+# Spice.ai Extensibility
+
+This document is an overview of all the interfaces and extension points in Spice.ai.
+
+## DataSource
```
Comment (lukekim) [inline docs/EXTENSIBILITY.md]: Maybe we should continue to call these `DataConnectors` especially if they now include writes.

File: docs/EXTENSIBILITY.md [hunk 4]
```diff
@@ -0,0 +1,82 @@
...
+/// If `stream_data_updates` is not supported for a dataset, the runtime will fall back to polling `get_all_data` and returning a
+/// `DataUpdate` that is constructed like:
+///
+/// ```rust
+/// DataUpdate {
+///    data: get_all_data(dataset),
+///    update_type: UpdateType::Overwrite,
+/// }
+/// ```
+pub trait DataConnector: Send + Sync {
```
Comment (mitchdevenport) [inline docs/EXTENSIBILITY.md]: I'm not a fan of documenting code in a separate location from the code. It gets hard to remember to update both locations and they end up diverging. How about we list the interfaces and extensions here with a quick summary of what they are and then link to the code itself where we full document?

**Review by mitchdevenport — DISMISSED** [review]

File: bin/spiced/src/lib.rs [hunk 5]
```diff
@@ -19,31 +19,31 @@ pub enum Error {
...
     },
 
     #[snafu(display("Unable to attach view"))]
     UnableToAttachView { source: runtime::datafusion::Error },
 
-    #[snafu(display("Unable to initialize data source: {data_source}"))]
-    UnableToInitializeDataSource {
-        source: runtime::datasource::Error,
-        data_source: String,
+    #[snafu(display("Unable to initialize data source: {data_connector}"))]
```
Comment (lukekim) [inline bin/spiced/src/lib.rs]:
```suggestion
    #[snafu(display("Unable to initialize data connector: {data_connector}"))]
```

File: bin/spiced/src/lib.rs [hunk 6]
```diff
@@ -19,31 +19,31 @@ pub enum Error {
...
     ))]
     RequiredParameterMissing {
         parameter: &'static str,
-        data_source: String,
+        data_connector: String,
     },
 
-    #[snafu(display("Unknown data source: {data_source}"))]
-    UnknownDataSource { data_source: String },
+    #[snafu(display("Unknown data source: {data_connector}"))]
```
Comment (lukekim) [inline bin/spiced/src/lib.rs]:
```suggestion
    #[snafu(display("Unknown data connector: {data_connector}"))]
```

File: docs/EXTENSIBILITY.md (hunk 4 above)
Comment (lukekim) [reply docs/EXTENSIBILITY.md]: I somewhat agree. I would use table format with links.

File: docs/EXTENSIBILITY.md [hunk 7]
```diff
@@ -0,0 +1,82 @@
+# Spice.ai Extensibility
+
+This document is an overview of all the interfaces and extension points in Spice.ai.
+
+## DataConnector
+
+A `DataConnector` is a component that represents the source of data to the Spice.ai runtime. Datasets specify which `DataConnector` to use via the `from` field in the dataset configuration. A `DataConnector` knows how to retrieve data, stream data updates, and writing data back to the source.
```
Comment (lukekim) [inline docs/EXTENSIBILITY.md:7]: Definitions should just be in `github.com/spiceai/docs`

**Review by lukekim — APPROVED** [review]

**Pull request merged**